lib/dynamodb/dynamodb_request_queue.c

```diff
@@ -243,6 +221,63 @@ extractcapacity(struct http_response * res, double * pcap)
 	return (-1);
 }
 
+/* Extract a DynamoDB error string from an HTTP 400 response. */
+static int
+extracterror(struct http_response * res, char ** errstr)
+{
+	const uint8_t * buf = res->body;
+	const uint8_t * end = res->body + res->bodylen;
+	size_t len;
+
+	/* Look for __type. */
+	buf = json_find(buf, end, "__type");
+
+	/*
+	 * We should be pointing at the opening double-quote of a string;
+	 * move past it.
+	 */
+	if (buf[0] != '"')
+		goto bad;
+	buf++;
+
+	/* Find a '#' inside the string and move past it. */
+	while (buf[0] != '#') {
+		if (buf[0] == '"')
+			goto bad;
+		buf++;
+	}
+	buf++;
+
+	/* The error string is everything from here to the closing '"'. */
+	for (len = 0; &buf[len] < end; len++) {
+		if (buf[len] == '"')
+			break;
+	}
+	if (&buf[len] == end)
+		goto bad;
+
+	/* Duplicate the error string. */
+	if ((*errstr = malloc(len + 1)) == NULL)
+		goto err0;
+	memcpy(*errstr, buf, len);
+	(*errstr)[len] = '\0';
+
+	/* Success! */
+	return (0);
+
+bad:
+	warn0("DynamoDB sent HTTP 400 response without valid error string!");
+	if ((*errstr = strdup("ErrorNotSpecified")) == NULL)
```

Comment:
... `strdup` works here, after all?

Ah, ok, granted `strndup` was added to POSIX in 2008, whereas `strdup` was 2001.  Was that the reason?